contracts/libraries/VaultLifecycleTreasury.sol

```diff
@@ -0,0 +1,796 @@
+import {SupportsNonCompliantERC20} from "./SupportsNonCompliantERC20.sol";
+import {UniswapRouter} from "./UniswapRouter.sol";
+
+library VaultLifecycleTreasury {
+    using SafeMath for uint256;
+    using SupportsNonCompliantERC20 for IERC20;
+
+    struct CloseParams {
+        address OTOKEN_FACTORY;
+        address USDC;
+        address currentOption;
+        uint256 delay;
+        uint16 lastStrikeOverrideRound;
+        uint256 overriddenStrikePrice;
+    }
+
+    /**
+     * @notice Sets the next option the vault will be shorting, and calculates its premium for the auction
+     * @param strikeSelection is the address of the contract with strike selection logic
+     * @param optionsPremiumPricer is the address of the contract with the
+       black-scholes premium calculation logic
+     * @param premiumDiscount is the vault's discount applied to the premium
+     * @param closeParams is the struct with details on previous option and strike selection details
+     * @param vaultParams is the struct with vault general data
+     * @param vaultState is the struct with vault accounting state
+     * @return otokenAddress is the address of the new option
+     * @return premium is the premium of the new option
+     * @return strikePrice is the strike price of the new option
+     * @return delta is the delta of the new option
+     */
+    function commitAndClose(
+        address strikeSelection,
+        address optionsPremiumPricer,
+        uint256 premiumDiscount,
+        CloseParams calldata closeParams,
+        Vault.VaultParams storage vaultParams,
+        Vault.VaultState storage vaultState
+    )
+        external
+        returns (
+            address otokenAddress,
+            uint256 premium,
+            uint256 strikePrice,
+            uint256 delta
+        )
+    {
+        uint256 expiry;
+
+        // uninitialized state
+        if (closeParams.currentOption == address(0)) {
+            expiry = getNextFriday(block.timestamp);
+        } else {
+            expiry = getNextFriday(
+                IOtoken(closeParams.currentOption).expiryTimestamp()
+            );
+        }
+
+        IStrikeSelection selection = IStrikeSelection(strikeSelection);
+
+        bool isPut = vaultParams.isPut;
+        address underlying = vaultParams.underlying;
+        address asset = vaultParams.asset;
+
+        (strikePrice, delta) = closeParams.lastStrikeOverrideRound ==
+            vaultState.round
+            ? (closeParams.overriddenStrikePrice, selection.delta())
+            : selection.getStrikePrice(expiry, isPut);
+
+        require(strikePrice != 0, "!strikePrice");
+
+        // retrieve address if option already exists, or deploy it
+        otokenAddress = getOrDeployOtoken(
+            closeParams,
+            vaultParams,
+            underlying,
+            asset,
+            strikePrice,
+            expiry,
+            isPut
+        );
+
+        // get the black scholes premium of the option
```

Comment:
No need for this comment.  Maybe we need to rename the function to `getOptionPremium` or `getContractPrice`?

Reply:
Im re-using this from existing VaultlifeCycle